File: 1474-longest-zigzag-path-in-a-binary-tree/1474-longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:

        self.maxZigzagLength = 0
        
        def dfs(node: Optional[TreeNode], move: str, currLength: int):
            if not node:
                return

            self.maxZigzagLength = max(self.maxZigzagLength, currLength)

            if move == 'L':
                dfs(node.right, 'R', currLength + 1)
                dfs(node.left, 'L', 1)

            elif move == 'R':
                dfs(node.left, 'L', currLength + 1)
                dfs(node.right, 'R', 1)

        dfs(root.left, 'L', 1)
        dfs(root.right, 'R', 1)
        return self.maxZigzagLength

File: 1474-longest-zigzag-path-in-a-binary-tree/1474-longest-zigzag-path-in-a-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function longestZigZag(root: TreeNode | null): number {
    let maxZigzagLength = 0

    const dfs = (node: TreeNode | null, move: string, currLength: number) => {
        if (!node) return

        maxZigzagLength = Math.max(maxZigzagLength, currLength)

        if (move === 'L') {
            dfs(node.right, 'R', currLength + 1)
            dfs(node.left, 'L', 1)
        }

        else if (move === 'R') {
            dfs(node.left, 'L', currLength + 1)
            dfs(node.right, 'R', 1)
        }
    }

    dfs(root.left, 'L', 1)
    dfs(root.right, 'R', 1)
    return maxZigzagLength
};